### DISCUSSION TITLE
ACE on Silicon DFT data

### DISCUSSION THREAD

**Discussion by exapde** [ACEsuit/ACE1.jl#27] [Help with ACE Fits]
Hi!

I used ACE to fit the Silicon DFT data from this paper https://arxiv.org/pdf/1906.08888.pdf. The data can be downloaded from the following Github site https://github.com/materialsvirtuallab/mlearn/tree/master/data.

The results are really strange because the errors in the energy and forces for the ACE potential is 10 times more than those of the SNAP potential.

Any help would be appreciated.

Thanks,
Cuong

**Comment by cortner** [comment]
Thanks for checking. Can you post the script you used to generate the basis and for the parameter estimation?

**Reply by cortner** [reply]
@casv2 would you be any chance be willing to follow along?

**Reply by exapde** [reply]
r0 = rnn(:Si)
        basis = rpi_basis(; 
              species = :Si,
              N = 3,                        # correlation order = body-order - 1
              maxdeg = 12,                  # polynomial degree
              D = SparsePSHDegree(; wL=1.5, csp=1.5),
              r0 = r0,                      # estimate for NN distance
              rin = 0.65*r0, rcut = 5.0,    # domain for radial basis (cf documentation)
              pin = 0)

**Reply by exapde** [reply]
Are the parameters I used okay?

**Reply by cortner** [reply]
A priori yes, you can play with degree and N = 3 might be too high for this dataset. Are you checking the training error or the test error?

**Reply by exapde** [reply]
I checked both. They are similar.

**Reply by cortner** [reply]
And does SNAP use the same cutoff?

**Reply by exapde** [reply]
I used the same cut-off for both of them.

**Reply by cortner** [reply]
and what is `maxn, maxl`  for SNAP

**Comment by cortner** [comment]
What is the parameter estimation? How do you regularise?

**Reply by exapde** [reply]
I used the same weights for both ACE and SNAP. Energy weight = 100, force weight = 1.

**Reply by cortner** [reply]
But how do you regularise?

**Reply by exapde** [reply]
I don't use any regularization.

**Reply by exapde** [reply]
How do you regularize ACE?

**Reply by cortner** [reply]
Ok thats interesting. I’ll have to look into it then.

**Comment by cortner** [comment]
If you’d be willing to post your complete script then I’ll try to reproduce your results?

**Reply by exapde** [reply]
It will take time for me to prepare the complete script. I'll get back to you when I have it ready for you.

**Reply by exapde** [reply]
I used the ZBL potential as a reference potential for both SNAP and ACE, since this is typical of SNAP potential.  However, this may not be necessary for ACE. After removing the ZBL potential for ACE, the results are improved. ACE's energy errors are 4 times larger than SNAP's,  and ACE's force errors are 3 times larger than SNAP's.

**Reply by cortner** [reply]
ZBL is very poor around 2.3A, which is why we almost never use it. But a factor 3 is still too much.

**Reply by cortner** [reply]
we will figure it out, I'm sure...

**Comment by casv2** [comment]
I would assume your outer cutoffs are identical to SNAP (`rcut`)? Also, just to be sure, there is no data with bonds smaller than `rin=0.65*r0`, right? If decreasing the `rin` changes the errors that means you're not fully describing all the data with the ACE.

The next step would be to increase polynomial degree before increasing `N`. I've had success with `N=4` and have gone up to polynomial degree `deg=19` or perhaps even higher to achieve ~1 meV/atom accuracies.  

Also what is your r0? 2.35126?

**Reply by cortner** [reply]
you normally need to look at the radial distribution function and make sure that there is no data below the inner cutoff - but since you chose pin = 0 this really should not make a bit difference. Because you are trying to reproduce SNAP I would actually go down to N = 2 and go up with the degree.

**Reply by cortner** [reply]
regarding the outer cutoff - you need to choose it of course, I don't know a priori what a good cutoff is for this specific training set?  Your r0 is fine and anyhow this makes very little difference.

**Reply by casv2** [reply]
Good point, there's generally a lot of accuracy to be gained in any `N` by upping the polynomial degree from say 12 to 16 or even higher

**Reply by cortner** [reply]
So if SNAP uses an rcut of 7A then that might partially explain the result.

**Reply by exapde** [reply]
I use 5A cut-off for both SNAP and ACE.

**Reply by cortner** [reply]
but the polynomial degree is more likely the culprit. Note we use sparse polynomials. So even if SNAP uses `maxn, maxl` fairly moderate these parameters don't mean the same for ACE.

**Reply by exapde** [reply]
SNAP has only one parameter 2J_max to control the number of descriptors.

**Reply by exapde** [reply]
Yes, my r0 is 2.35126

**Reply by cortner** [reply]
then a simple way to compare would be to try and adjust the polynomial degree to have a similar number of features overall. The other useful piece of information about SNAP would be how they balance l and n, but we can get there afterwards.

**Reply by exapde** [reply]
Can ACE handle non-orthogonal lattices, for example, like the one below

      a = [14.22366 5.591921 15.557314]
      b = [0.0 9.055935 0.272179]
      c = [0.0 0.0 9.178194]

There are many structures like this in the data I used.

**Reply by cortner** [reply]
yes - this is entirely independent of ACE and implemented in JuLIP.jl and Neighbourlists.jl which are thoroughly tested against ase and against James Kermode's and Lars Pastewka's implementations.

**Reply by exapde** [reply]
Good news! After transpose the lattices in my data, ACE's errors are similar to SNAP.

**Reply by exapde** [reply]
How the lattice vectors in Atoms.cell are defined? Is it by rows or columns?

**Reply by cortner** [reply]
I think it is the transpose of the ASE definition, but that's all I remember. I use `pymatgen` to load the data, then convert to ASE, then to `ASE.jl` then to `JuLIP.jl` - I can send you a script if that would help?

**Reply by exapde** [reply]
I think I did not understand the notation of lattices in ACE's Atoms struct.  I thought that the lattice vectors are columns of Atoms.cell. It seems that they are rows of Atoms.cell

**Reply by cortner** [reply]
Does this help:
```
julia> using JuLIP
[ Info: Precompiling JuLIP [945c410c-986d-556a-acb1-167a618e0462]

julia> cell(bulk(:Ti))
3×3 StaticArrays.SMatrix{3, 3, Float64, 9} with indices SOneTo(3)×SOneTo(3):
  2.95   0.0      0.0
 -1.475  2.55477  0.0
  0.0    0.0      4.6846
```

**Reply by cortner** [reply]
Here is what I'm using to load that data; this should be correct.
```Julia

using PyCall, ASE, JuLIP 

ase = pyimport("ase")
pmt = pyimport("pymatgen.core")
pmt2ase = pyimport("pymatgen.io.ase")
json = pyimport("json")

f = open(@__DIR__() * "/Si_training.json", "r")
data = json.load(f)
train = []
for d in data
   curr_structure = pmt.Structure.from_dict(d["structure"])
   at_ase = pmt2ase.AseAtomsAdaptor.get_atoms(curr_structure)
   at = Atoms(ASEAtoms(at_ase))
   push!(train, at)
end
```
Haven't had time though to test the ACE basis on those.

**Reply by cortner** [reply]
there is an issue to put this into ACEsuit, but nobody has taken that up yet :)

**Reply by exapde** [reply]
Thank you very much. You are super helpful.

**Reply by cortner** [reply]
updated script: 
```Julia


using PyCall, ASE, JuLIP, ACE1pack.IPFitting 
import ACE1pack.IPFitting: Dat

ase = pyimport("ase")
pmt = pyimport("pymatgen.core")
pmt2ase = pyimport("pymatgen.io.ase")
json = pyimport("json")

f = open(@__DIR__() * "/Si_training.json", "r")
data = json.load(f)
train = []
for d in data
   curr_structure = pmt.Structure.from_dict(d["structure"])
   at_ase = pmt2ase.AseAtomsAdaptor.get_atoms(curr_structure)
   at = Atoms(ASEAtoms(at_ase))
   dat = Dat(at, "all", E = d["outputs"]["energy"], 
                        F = vecs(d["outputs"]["forces"]'),  )
   push!(train, dat)
end

train = identity.(train)
```

How is the virial stored (6 components instead of 9 - do you know how to convert that so we can add it to the fit?)

**Reply by exapde** [reply]
I don't use virial since, in the paper, they only used energy and force for fitting and testing.

**Comment by cortner** [comment]
Note, btw, that we are now working towards removing all this freedom in the parameter choices and fully automating it. It must look awful right now, and it will get a lot better :).

**Reply by exapde** [reply]
That's really good.

**Comment by cortner** [comment]
Ok - I have a script that is running, what is your target accuracy for energy and forces? Do you use virials?

**Reply by cortner** [reply]
here is my first naive attempt
```
┌─────────────┬─────────┬──────────┬────────┬────────┐
│ config type │ E [meV] │ F [eV/A] │  F [%] │ V[meV] │
├─────────────┼─────────┼──────────┼────────┼────────┤
│         all │   5.243 │    0.252 │ 28.888 │    NaN │
├─────────────┼─────────┼──────────┼────────┼────────┤
│         set │   5.243 │    0.252 │ 28.888 │    NaN │
└─────────────┴─────────┴──────────┴────────┴────────┘

```

**Reply by cortner** [reply]
far from great - I'm clearly missing something important.

**Reply by cortner** [reply]
with `N = 3`: 
```
┌─────────────┬─────────┬──────────┬────────┬────────┐
│ config type │ E [meV] │ F [eV/A] │  F [%] │ V[meV] │
├─────────────┼─────────┼──────────┼────────┼────────┤
│         all │   0.979 │    0.118 │ 13.541 │    NaN │
├─────────────┼─────────┼──────────┼────────┼────────┤
│         set │   0.979 │    0.118 │ 13.541 │    NaN │
└─────────────┴─────────┴──────────┴────────┴────────┘
```

**Reply by cortner** [reply]
note these are training errors, not test errors. But comparing to the manuscript this looks promising. Now would need to fine-tune the regularization. but I think you should be able to reach the accuracy of SNAP now.

**Reply by cortner** [reply]
Here is the full script: 
```Julia
using ACE1pack, JuLIP 
using ACE1pack.IPFitting

include("Si.jl")

##

r0 = rnn(:Si)

basis = rpi_basis(; 
      species = :Si,
      N = 3,                        # correlation order = body-order - 1
      maxdeg = 16,                  # polynomial degree
      r0 = r0,                      # estimate for NN distance
      D = SparsePSHDegree(; wL=1.3, csp=1.0),
      rin = 0.65*r0, rcut = 5.0,    # domain for radial basis (cf documentation)
      pin = 0)

@show length(basis)


dB = LsqDB("", basis, train)

weights = Dict("default" => Dict("E" => 100.0, "F" => 1.0 , "V" => 1.0 ))

IP, lsqinfo = lsqfit(dB; weights = weights, asmerrs = false)

ACE1pack.IPFitting.add_fits!(IP, train)

ACE1pack.IPFitting.rmse_table(train)
```

and `Si.jl`: 
```Julia
using PyCall, ASE, JuLIP, ACE1pack.IPFitting 
import ACE1pack.IPFitting: Dat

ase = pyimport("ase")
pmt = pyimport("pymatgen.core")
pmt2ase = pyimport("pymatgen.io.ase")
json = pyimport("json")

f = open(@__DIR__() * "/Si_training.json", "r")
data = json.load(f)
train = []
for d in data
   curr_structure = pmt.Structure.from_dict(d["structure"])
   at_ase = pmt2ase.AseAtomsAdaptor.get_atoms(curr_structure)
   at = Atoms(ASEAtoms(at_ase))
   dat = Dat(at, "all", E = d["outputs"]["energy"], 
                        F = vecs(d["outputs"]["forces"]'),  )
   push!(train, dat)
end

train = identity.(train)
```

and finally my `Project.toml`:
```toml
[deps]
ACE1pack = "8c4e8d19-0bd6-4234-8309-7210652e3178"
ASE = "51974c44-a7ed-5088-b8be-3e78c8ba416c"
JSON = "682c06a0-de6a-54ab-a142-c8b1cf79cde6"
JuLIP = "945c410c-986d-556a-acb1-167a618e0462"
PyCall = "438e738f-606a-5dbb-bf0a-cddfbfd45ab0"
```

**Reply by cortner** [reply]
If you are happy with the training accuracy but can't get good test accuracy, then please let's continue the discussion here and I'll help you sort this out too. Basically you need to tell IPFitting how to rescale the basis and then which form of regularization to use.

**Reply by exapde** [reply]
Yes, I got very similar errors! Thanks very much.

**Comment by cortner** [comment]
Here is an example script with regularization - I can just and barely reach SNAP accuracy now. But note this is with essentially no fine-tuning! To do better we need to play around optimizing the basis a bit and especially finding just the right sparsification and regularization. Maybe a quick thing to try immediately would be for @casv2 to point you to an example how to use BRR (Baysian ridge regression) which he has had huge success with! 
```Julia

using ACE1pack, JuLIP 
using ACE1pack.IPFitting
using LinearAlgebra

include("Si.jl")

##

r0 = rnn(:Si)

basis = rpi_basis(; 
      species = :Si,
      N = 3,                        # correlation order = body-order - 1
      maxdeg = 18,                  # polynomial degree
      r0 = r0,                      # estimate for NN distance
      D = SparsePSHDegree(; wL=1.3, csp=1.0),
      rin = 0.65*r0, rcut = 5.0,    # domain for radial basis (cf documentation)
      pin = 0)

@show length(basis)


dB = LsqDB("", basis, train)

weights = Dict("default" => Dict("E" => 100.0, "F" => 1.0 , "V" => 1.0 ))

reg = Diagonal(ACE1.scaling(basis, 2))

IP, lsqinfo = lsqfit(dB; weights = weights, asmerrs = false, 
                        solver = Dict("solver" => :rrqr, "rrqr_tol" => 1e-6, 
                                      "P" => reg))
                     #   solver = Dict("solver" => :brr, "brr_tol" => 1e-5))

ACE1pack.IPFitting.add_fits!(IP, train)
ACE1pack.IPFitting.add_fits!(IP, test)
ACE1pack.IPFitting.rmse_table(train)
ACE1pack.IPFitting.rmse_table(test)
```

**Reply by cortner** [reply]
```
TRAINING ERROR 
┌─────────────┬─────────┬──────────┬────────┬────────┐
│ config type │ E [meV] │ F [eV/A] │  F [%] │ V[meV] │
├─────────────┼─────────┼──────────┼────────┼────────┤
│         all │   1.359 │    0.133 │ 15.278 │    NaN │
├─────────────┼─────────┼──────────┼────────┼────────┤
│         set │   1.359 │    0.133 │ 15.278 │    NaN │
└─────────────┴─────────┴──────────┴────────┴────────┘

TEST ERROR
┌─────────────┬─────────┬──────────┬────────┬────────┐
│ config type │ E [meV] │ F [eV/A] │  F [%] │ V[meV] │
├─────────────┼─────────┼──────────┼────────┼────────┤
│         all │   6.302 │    0.159 │ 18.073 │    NaN │
├─────────────┼─────────┼──────────┼────────┼────────┤
│         set │   6.302 │    0.159 │ 18.073 │    NaN │
└─────────────┴─────────┴──────────┴────────┴────────┘
```

**Reply by exapde** [reply]
I tried to optimize both rin and rcut. And it helps reduce the errors a little bit. I will be happy to try BRR.

**Comment by cortner** [comment]
<img width="844" alt="Screen Shot 2022-02-23 at 9 34 48 AM" src="https://user-images.githubusercontent.com/7688515/155375461-c9d07027-e374-4fde-9281-43a091117362.png">

Just saw this in a seminar - a different training set from the same paper. They manage to reach full accuracy with standard linear ACE. We need to figure out how to achieve this automatically without fiddling.

**Reply by casv2** [reply]
I'm not that surpised, I think I'm very much on that line for many of these ACE potentials I'm using in NS. Model selection and ARD is what I use. Model selection to make sure you're not making the basis unnecesarily large, ARD to sparsify. I use generally ~10K basis functions but that's for 5 elements, not just a single one.

**Reply by casv2** [reply]
Actually wait a minute that's wrong, here's HfNbTaTiZr:

```[ Info: Fit complete: keeping 1987 basis functions (16.0%)```
```
┌────────────────┬─────────┬──────────┬────────┬────────┐
│    config type │ E [meV] │ F [eV/A] │  F [%] │ V[meV] │
├────────────────┼─────────┼──────────┼────────┼────────┤
│            set │   3.515 │    0.291 │  8.227 │ 76.252 │
└────────────────┴─────────┴──────────┴────────┴────────┘
```

Training errors though, but ARD generalises incredibly well the test set errors will not be far off. Also errors sure are a tad high, and this is very heavily sparsified. More importantly stuff like repulsiveness of the 2B work well here too

**Reply by cortner** [reply]
very nice. I'll need to try our your model selection. Unfortunately I can't install it on my M1 it seems...

**Reply by exapde** [reply]
Thank you for sharing the interesting results. I am curious about the error in forces as well. Do they have similar comparison for forces? If we try to optimize the errors in energy, then the errors in forces can be large, and vice versa. I think it is important to keep balance in fitting energy and forces.

**Reply by exapde** [reply]
The weights are optimized for training linear ACE potentials. I think they use genetic algorithm to optimize the weights, which is very computationally intensive.

**Reply by cortner** [reply]
but I think still cheap compared to the parameter estimation for nonlinear models... we are starting our very first tests with that and even on really small training sets and small models where I can do the linear fit on the order of seconds to minutes on my laptop it takes several hours on a workstation with 64 cores.

**Reply by exapde** [reply]
Comparing ACE and SNAP on the datasets in that paper, ACE and SNAP are similar in accuracy of predicted energy and forces when the number of descriptors is chosen the same. ACE is many times faster than SNAP for the same number of descriptors.

**Reply by cortner** [reply]
Interesting - I would have expected the opposite - that ACE can be easily made more accurate but that SNAP should be faster. The point is that SNAP has a much simpler structure which is much easier to optimise. But I guess we did some good things in our ACE code :)
